Sources/NIOHTTP1/NIOTypedHTTPServerUpgraderStateMachine.swift

```diff
@@ -393,5 +403,49 @@ struct NIOTypedHTTPServerUpgraderStateMachine<UpgradeResult> {
         }
     }
 
+    @usableFromInline
+    enum CloseInboundAction {
+        case close
+        case `continue`
+    }
+
+    @inlinable
+    mutating func closeInbound() -> CloseInboundAction {
+        switch self.state {
+        case .initial:
+            self.state = .finished
+            return .close
+
+        case .awaitingUpgrader(var awaitingUpgrader):
+            if awaitingUpgrader.seenFirstRequest {
+                // We should buffer the input close since we have seen the full request.
+                awaitingUpgrader.buffer.append(.inputClosed)
+                self.state = .awaitingUpgrader(awaitingUpgrader)
+                return .continue
+            } else {
+                // We shouldn't buffer. This means we were still expecting HTTP parts.
+                return .close
+            }
+
+        case .upgrading(var upgrading):
+            upgrading.buffer.append(.inputClosed)
+            self.state = .upgrading(upgrading)
+            return .continue
+
+        case .upgraderReady(var upgraderReady):
+            upgraderReady.buffer.append(.inputClosed)
+            self.state = .upgraderReady(upgraderReady)
+            return .continue
+
+        case .unbuffering(var unbuffering):
+            unbuffering.buffer.append(.inputClosed)
+            self.state = .unbuffering(unbuffering)
+            return .continue
+
+        case .modifying, .finished:
+            return .continue
```

Comment:
In finished we should forward the `inputClosed` event. In `modifying` we should `fatalError` as we do in the other methods.

Reply:
done